### DISCUSSION TITLE
Arduino C++ Coding

### DISCUSSION THREAD

**Discussion by JustinFans** [community/community#122575] [Programming Help]
### Body

Hi :)
I have a question that is how do I code in C++ on for example an Arduino Microcontroller like this one ELEGOO MEGA R3 Board ATmega 2560, because I downloaded the Arduino IDE and tried to code. However it looks like a different coding language like regular C or something, the problem is I have to code in C++ and the other coding languages don't count for my task. I couldn't find an answer when I was searching online and I know people here are extremely talented in coding and things like that I hope someone can help me out. :D 
kind regards,
Justin

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Pigamer37** [comment]
You should be able to program in regular C/C++. The arduino library includes methods/functions/etc. to interact with the hardware, and substitutes the classic `main` function as the program's entry point with a `setup` (run once before looping) and `loop` (the main logic of your program, looping forever) functions. Otherwise, the syntax should be the exact same.

Hope this helps!

**Comment by alokkcodex** [comment]
Pigamer37's right — you're already writing C++, it just doesn't look like the C++ you might be used to (e.g. from a Windows/Linux/Mac tutorial) because of two things:

1. **No `main()` function.** Arduino replaces it with:
   - `void setup() { }` — runs once when the board powers on/resets
   - `void loop() { }` — runs repeatedly after that

   Under the hood, the Arduino core actually has its own `main()` that calls `setup()` once and then `loop()` forever — you just never see it.

2. **The Arduino "language" is really just C++ plus a bundled library.** Functions like `digitalWrite()`, `pinMode()`, `Serial.print()`, etc. aren't new syntax — they're regular C++ functions/classes defined in the Arduino core library, which gets auto-included for you.

Everything else — classes, structs, pointers, `#include`, templates, the STL (with some caveats on AVR boards like the Mega 2560, which has limited memory so heavy STL use like `<vector>` can be risky), operator overloading, etc. — is standard C++ and works the same way.

So if your task requires "C++" specifically: yes, Arduino sketches satisfy that. If a professor/grader is strict about seeing a `main()` function or wants it built outside the Arduino IDE, you can also compile Arduino-style code with a plain C++ toolchain (e.g. PlatformIO or avr-gcc directly) and write your own `main()` — but for learning/prototyping, the IDE's `setup()`/`loop()` model is just C++ with training wheels, not a different language.

**Comment by tuna4ll** [comment]
Arduino sketches are already compiled as C++, even though the syntax may look simpler at first.

The Arduino IDE automatically generates a few things behind the scenes (such as function prototypes and the `main()` entry point), which is why it doesn't look like a typical C++ project. However, you can freely use standard C++ features such as:

* classes and objects
* inheritance
* templates (within the limits of the compiler and available memory)
* namespaces
* operator overloading
* `.cpp` and `.h` files

For example, this is perfectly valid Arduino code:

```cpp
class Led {
public:
    explicit Led(int pin) : pin(pin) {}

    void begin() {
        pinMode(pin, OUTPUT);
    }

    void on() {
        digitalWrite(pin, HIGH);
    }

    void off() {
        digitalWrite(pin, LOW);
    }

private:
    int pin;
};

Led led(13);

void setup() {
    led.begin();
}

void loop() {
    led.on();
    delay(500);
    led.off();
    delay(500);
}
```

So if your assignment requires C++, Arduino satisfies that requirement. The only caveat is that not every part of the C++ standard library is available on a microcontroller due to memory and hardware limitations.
